### PR TITLE
[Android, WPF] Fixes rounding down the size of visual elements

### LINKED ISSUE #1544: StackLayout with zero spacing is not always zero spacing
### Description

A `StackLayout` with `Spacing="0"` will not always honor that spacing if it contains controls of certain heights.

### Steps to Reproduce

```XML
    <StackLayout Spacing="0">
        <Label
            HeightRequest="55.7"/>

        <BoxView
            BackgroundColor="#433DBA"
            HeightRequest="10"/>
        <BoxView
            BackgroundColor="#6461B7"
            HeightRequest="10"/>
        <BoxView
            BackgroundColor="#433DBA"
            HeightRequest="10"/>
        <BoxView
            BackgroundColor="#6461B7"
            HeightRequest="10"/>
        ...
    </StackLayout>
```

### Expected Behavior

Each `BoxView` should be right up against the previous `BoxView` with no gaps.

### Actual Behavior

There are gaps periodically in the children following the `Label` with height `55.7` (I am sure there are other heights that trigger the same behavior, but I didn't bother looking for them).

### Basic Information

- Version with issue: `2.5.0.121934`
- Last known good version: N/A
- IDE: VS
- Platform Target Frameworks:
  - Android: 8.0

### Screenshots

screenshot_20180110-102148

### Reproduction Link

[XFTestBed.zip](https://github.com/xamarin/Xamarin.Forms/files/1617115/XFTestBed.zip)

### PR DESCRIPTION
**Pull request by paymicro**
### Description of Change ###

Fixes rounding down the size of visual elements

<details>
<summary>Before</summary>

screenshot_1526296524

untitled2

The height of the elements is distorted if it is not a multiple of 0.8. 
(10 `Height` = 9.6 `ActualHeight` | 10.1 `Height`=> 10.4 `ActualHeight`)

</details>

<details>
<summary>After</summary>

screenshot_1526296206

untitled
</details>

### Bugs Fixed ###

Fixes #1544

### API Changes ###

/

### Behavioral Changes ###

- In `WPF` and `Android`, the actual height is only rounded up.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/VisualElement.cs [hunk 1]
```diff
@@ -573,9 +574,36 @@ public virtual SizeRequest GetSizeRequest(double widthConstraint, double heightC
 			return r;
 		}
 
+		// In WPF and Android, the actual height is rounded in step of 0.8 points.
+		// So that the real size was no less than the preset it is necessary to increase it so that the rounding was only in the larger side.
+		Rectangle PrepareForRounding(Rectangle region)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/VisualElement.cs]: shouldn't this be in platform code ?

**Comment by mohachouch** [comment]
@paymicro Can you resolved the conflict file.  Thks

Why is it just needed for Android and WPF? (UWP is very similar to WPF)

**Comment by paymicro** [comment]
@mohachouch on scaled monitor WPF rounds the view size to nearest step. 
UWP always rounds up to next step. Therefore this fix need only for WPF and Android.
<details>
<summary>WPF before</summary>

screenshot_4
</details>
<details>
<summary>UWP before/after</summary>

screenshot_5
</details>

File: Xamarin.Forms.Core/VisualElement.cs [hunk 2]
```diff
@@ -679,9 +679,11 @@ public virtual SizeRequest GetSizeRequest(double widthConstraint, double heightC
 			return r;
 		}
 
+		public Func<Rectangle, Rectangle> PrepareLayout;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/VisualElement.cs]: 👎

File: Xamarin.Forms.Core/VisualElement.cs [hunk 3]
```diff
@@ -679,9 +679,11 @@ public virtual SizeRequest GetSizeRequest(double widthConstraint, double heightC
...
+
 		public void Layout(Rectangle bounds)
 		{
-			Bounds = bounds;
+			Bounds = PrepareLayout != null ? PrepareLayout(bounds) : bounds;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/VisualElement.cs]: should be part of the measuring calls

**Comment by samhouts** [comment]
@paymicro Can you address @StephaneDelcroix's comments? Thanks!

**Comment by samhouts** [comment]
@paymicro I want to merge this, but almost every test is failing on API28 Fast Renderers. Can you take a look and make sure nothing is broken? Thanks!

**Comment by paymicro** [comment]
@samhouts looks like the tests just didn't run at API28...

**Comment by rmarinho** [comment]
Api28 failure is not related

**Pull request merged**